toolchain/lower/mangler.cpp

```diff
@@ -0,0 +1,99 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/lower/mangler.h"
+
+#include "toolchain/base/kind_switch.h"
+#include "toolchain/sem_ir/entry_point.h"
+
+namespace Carbon::Lower {
+
+auto Mangler::MangleInverseQualifiedNameScope(llvm::raw_ostream& os,
+                                              SemIR::NameScopeId name_scope_id)
+    -> void {
+  struct NameEntry {
+    SemIR::NameScopeId name_id;
+    char prefix;
+  };
+  llvm::SmallVector<NameEntry> names_to_render;
+  auto add_scope = [&](SemIR::NameScopeId name_scope_id, char prefix) {
+    if (name_scope_id.is_valid() &&
+        name_scope_id != SemIR::NameScopeId::Package) {
+      names_to_render.push_back({name_scope_id, prefix});
+    }
+  };
+  add_scope(name_scope_id, '.');
+  while (!names_to_render.empty()) {
+    auto [name_scope_id, prefix] = names_to_render.back();
+    names_to_render.pop_back();
+    const auto& name_scope = sem_ir().name_scopes().Get(name_scope_id);
+    if (prefix) {
+      os << prefix;
+    }
+    add_scope(name_scope.parent_scope_id, '.');
+    CARBON_KIND_SWITCH(sem_ir().insts().Get(name_scope.inst_id)) {
+      case CARBON_KIND(SemIR::ImplDecl impl_decl): {
+        const auto& impl = sem_ir().impls().Get(impl_decl.impl_id);
+
+        auto interface_type =
+            types().GetAs<SemIR::InterfaceType>(impl.constraint_id);
+        auto interface = sem_ir().interfaces().Get(interface_type.interface_id);
+        names_to_render.push_back({interface.scope_id, ':'});
```

Comment:
Can you use `add_scope` where you currently use `names_to_render.push_back`? Mainly thinking for consistency, one way other the other.

Reply:
I had been using direct push_back when the extra conditionals (valid/package) weren't needed, but now that those conditionals aren't needed at all - I removed the helper entirely and left these all as raw push_back, because that seemed still sufficiently legible to not warrant the extra lambda indirection to me, at least.